bpo-31368: Enhance os.preadv() documentation

vstinner opened the pull request:
<!-- issue-number: bpo-31368 -->
https://bugs.python.org/issue31368
<!-- /issue-number -->

vstinner committed:
bpo-31368: Enhance os.preadv() documentation

vstinner committed:
What's New: describe added functions

csabella commented on Doc/library/os.rst:
```diff
@@ -1104,39 +1104,48 @@ or `the MSDN <https://msdn.microsoft.com/en-us/library/z0kc8e3z.aspx>`_ on Windo
+   Write the *buffers* contents to file descriptor *fd* at offset *offset*.
+   *buffers* must be a sequence of :term:`bytes-like objects <bytes-like
+   object>`. Buffers are processed in array order. Entire contents of first
+   buffer is written before proceeding to second, and so on.
```
'Entire contents of ***the*** first buffer'
'before proceeding to **the** second'

vstinner replied:
done

vstinner committed:
Adding missing "the" in the doc

vstinner committed:
Move preadv to fix ordering

pablogsal commented on Doc/library/os.rst:
```diff
@@ -1328,9 +1350,10 @@ or `the MSDN <https://msdn.microsoft.com/en-us/library/z0kc8e3z.aspx>`_ on Windo
-   (sysconf() value SC_IOV_MAX) on the number of buffers that can be used.
+   processed in array order. Entire contents of the first buffer is written
+   before proceeding to the second, and so on. The operating system may set a
+   limit (sysconf() value SC_IOV_MAX) on the number of buffers that can be
```
It will make sense to change 
```
 sysconf() -> :func:`sysconf` 
```
and
```
SC_IOV_MAX ->  ``'SC_IOV_MAX'``
```
?

vstinner replied:
Oh wait, I didn't want to modify this documentation... But then I read the doc, and oh, all these documentations can be enhanced...

YoSTEALTH commented on Doc/library/os.rst:
```diff
@@ -1092,6 +1092,64 @@ or `the MSDN <https://msdn.microsoft.com/en-us/library/z0kc8e3z.aspx>`_ on Windo
+   - :data:`RWF_HIPRI`
+   - :data:`RWF_NOWAIT`
+
+   Using non-zero flags requires Linux 4.6 or newer.
```
could be change this to: "Using non-zero flags requires preadv2 support on Linux 4.7 or newer."

vstinner replied:
I'm trying to avoid mentioning preadv2(). Do you think that you must mention it?

YoSTEALTH replied:
All the flags actually belong to preadv2() function though, maybe a tiny mention wont hurt?
Its your call.

njsmith replied:
You should absolutely mention `preadv2` and `pwritev2`. At the point someone is using these features in the `os` module they're also reading man pages and lwn articles, and you want to help them make the connection. If someone wonders whether python has a wrapper for `preadv2` then searching for that in the docs should find `os.preadv`.

YoSTEALTH commented on Doc/library/os.rst:
```diff
@@ -1104,39 +1162,48 @@ or `the MSDN <https://msdn.microsoft.com/en-us/library/z0kc8e3z.aspx>`_ on Windo
+   - :data:`RWF_DSYNC`
+   - :data:`RWF_SYNC`
 
    Using non-zero flags requires Linux 4.7 or newer.
```
could be also change this to: "Using non-zero flags requires pwritev2 support on Linux 4.7 or newer."

vstinner committed:
Cleanup other read/write docs

pread(): rename second parameter from 'buffersize' to 'n', reuse
read() parameter name. The parameter is positional-only and so this
change is backward compatible.

YoSTEALTH commented on Doc/library/os.rst:
```diff
@@ -1092,6 +1092,64 @@ or `the MSDN <https://msdn.microsoft.com/en-us/library/z0kc8e3z.aspx>`_ on Windo
+   The operating system may set a limit (:func:`sysconf` value
+   ``'SC_IOV_MAX'``) on the number of buffers that can be used.
+
+   Availability: Linux 2.6.30 abd newer, FreeBSD 6.0 and newer,
```
should be "Availability: Linux 2.6.30, FreeBSD 6.0 and newer."

vstinner replied:
"and newer" is commonly used in os.rst documentation (i fixed the typo)

vadmium commented on Doc/library/os.rst:
```diff
@@ -1082,20 +1082,81 @@ or `the MSDN <https://msdn.microsoft.com/en-us/library/z0kc8e3z.aspx>`_ on Windo
+   The operating system may set a limit (:func:`sysconf` value
+   ``'SC_IOV_MAX'``) on the number of buffers that can be used.
+
+   Combine the functionality of :func:`os.readv` and :func:`os.pread`.
```
This reads as an instruction, but it is no longer clear it is for the implementation. Maybe drop the imperative and write “This function combines the functionality . . .”. Also applies to _pwritev_ below.

vadmium commented on Doc/library/os.rst:
```diff
@@ -1082,20 +1082,81 @@ or `the MSDN <https://msdn.microsoft.com/en-us/library/z0kc8e3z.aspx>`_ on Windo
+   - :data:`RWF_NOWAIT`
+
+   Return the total number of bytes actually read which can be less than the
+   total capacity of all the objects.
```
Needs a comma: “. . . number of bytes read, which can be less . . .”. Applies to _readv_ below too.

vadmium commented on Doc/library/os.rst:
```diff
@@ -1082,20 +1082,81 @@ or `the MSDN <https://msdn.microsoft.com/en-us/library/z0kc8e3z.aspx>`_ on Windo
+
+   If some data was successfully read, it will return the number of bytes read.
+   If no bytes were read, it will return ``-1`` and set errno to
+   :data:`errno.EAGAIN`.
```
I’m confused. Normally you access _errno_ as an attribute of an exception. Do the Python wrapper(s) return −1 or raise OSError/BlockingIOError?

vadmium commented on Doc/library/os.rst:
```diff
@@ -1104,48 +1165,57 @@ or `the MSDN <https://msdn.microsoft.com/en-us/library/z0kc8e3z.aspx>`_ on Windo
-   (sysconf() value SC_IOV_MAX) on the number of buffers that can be used.
-   :func:`~os.pwritev` writes the contents of each object to the file descriptor
-   and returns the total number of bytes written.
+   Write the *buffers* contents to file descriptor *fd* at a offset *offset*,
```
_a_ → _an_ or _the_, or even better “at offset _offset_” would work I think.
